go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,102 @@
+package libovsdbops
+
+import (
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+const (
+	// db ownership ExternalIDs
+	// when possible, use DbIndex and function from this file instead of directly using the following ExternalIDs
+	OwnerController = "owner_controller"
+	OwnerObjectType = "owner_object_type"
+	OwnerNamespace  = "owner_namespace"
+	OwnerName       = "owner_name"
+	InternalID      = "internal_id"
+
+	// Owner Object Types
+	EgressFirewallDNSOwnerType = "EgressFirewallDNS"
+	EgressQoSOwnerType         = "EgressQoS"
+	NetworkPolicyOwnerType     = "NetworkPolicy"
+	NamespaceOwnerType         = "Namespace"
+	HybridNodeRouteOwnerType   = "HybridNodeRoute"
+)
+
+type DbIndex struct {
+	// OwnerController specifies which controller owns the object
+	// Controller should only change objects it owns, make sure to always set this field when searching with predicate
+	OwnerController string
+	// OwnerObjectType should identify object type that owns db object, e.g. Pod, NetworkPolicy, etc.
+	// Controller handlers should use unique OwnerTypes to only change objects they manage.
+	OwnerObjectType string
+	// OwnerNamespace and OwnerName should provide unique index for objects owned by the same OwnerObjectType,
+	// e.g. if there is only 1 object owned by OwnerObjectType in the system, both Namespace and Name may be empty.
+	OwnerNamespace string
+	OwnerName      string
+	// InternalID may be used to additionally identify db objects owned by same OwnerNamespace and OwnerName.
+	InternalID string
+}
+
+func GetDbIndexFromExternalIDs(externalIDs map[string]string) *DbIndex {
+	return &DbIndex{
+		externalIDs[OwnerController],
+		externalIDs[OwnerObjectType],
+		externalIDs[OwnerNamespace],
+		externalIDs[OwnerName],
+		externalIDs[InternalID],
+	}
+}
+
+func (dbIndex *DbIndex) GetExternalIDs() map[string]string {
+	externalIDs := map[string]string{
+		OwnerController: dbIndex.OwnerController,
+		OwnerObjectType: dbIndex.OwnerObjectType,
+	}
+	if dbIndex.OwnerNamespace != "" {
+		externalIDs[OwnerNamespace] = dbIndex.OwnerNamespace
+	}
+	if dbIndex.OwnerName != "" {
+		externalIDs[OwnerName] = dbIndex.OwnerName
+	}
+	if dbIndex.InternalID != "" {
+		externalIDs[InternalID] = dbIndex.InternalID
+	}
+	return externalIDs
+}
+
+// GetObjName returns unique string that identifies object with given DbIndex
+func (dbIndex *DbIndex) GetObjName() string {
+	return dbIndex.OwnerController + "_" + dbIndex.OwnerObjectType + "_" + dbIndex.OwnerNamespace +
+		"_" + dbIndex.OwnerName + "_" + dbIndex.InternalID
```

Comment:
Good point, so Name is an indexed field for some tables (including address sets), and it must be unique. I intentionally left consecutive underscores, so that you can't end up with the same ids, e.g. if one object has only namespace set, and another one only name, but namespace and name are the same. Is we omit empty ids, these will end up being equal, but with not so pretty consecutive underscores it won't. Also it may be easier to read for humans, so that you always know if it's a namespace or name based on underscore position (e.g. may be useful for logging).

Reply:
In might be hard to the eyes, specially with some fonts, to tell apart one underscore from two consecutive underscores. So I would probably suggest a different separator.
It also looks kinda weird that for the same controller and object type (which are mandatory) we would end up in a mixed case where some of the other fields might be empty for some rows but not for others.

Reply:
right, so this DbIndex actually consists of 2 parts: the first part defines who manages given object, which is OwnerController and OwnerObjectType, and the second one identifies the object itself, it's OwnerNamespace, OwnerName, InternalID. 
The first part is required to communicate with the other "controllers" (any entity that creates/updates/deletes db objects). So it is required to be always set to make sure controllers don't change each others objects by accident. Then for object identification every controller can do what it wants, based on the object nature, e.g. if there is only 1 object owned by OwnerObjectType in the system, all the other ids may be empty.
For separators we can maybe use something like '/', or anything else except  '-' and '.'

Reply:
Understood. 

I still have a hard time thinking of a case where for the same OwnerController and OwnerObjectType, we would have different empty values for OwnerNamespace, OwnerName, InternalID. What I am thinking is that a given OwnerController and OwnerObjectType basically pins down which of OwnerNamespace, OwnerName, InternalID are required and which are not needed (but none are really optional once OwnerController and OwnerObjectType are set).

Anyway, it would be nice to have some more predictability and not that much flexibility but there might be a case where we might actually need for a controller to do whatever it wants with these ids, even weird things such as this.

Another separator we could use is `:` maybe?